tools/integration_tests_runner.py

```diff
@@ -1,167 +1,232 @@
+            )
+            for (test_file, test_name), (err, result) in test_results:
+                logger.info(
+                    err,
+                    extra={"file" : path.basename(test_file), "result" : result}
+                )
+
+        finally:
+            self.pool.close()
+            self.pool.join()
+
+
+    def map_args(self, test):
+        test_full_path = path.join(self.workspace_path, test)
+        test_in_exec_file = execute_file(test_full_path, "--list_tests")[0]  # a list
+
+        return map(
+            lambda tests_in_file: (test_full_path, tests_in_file, self.params),
+            test_in_exec_file
+        )
+
+
+    def prepare_list_of_tests(self):
+        for exec_file in self.runlist:
+            for test_tuple in self.map_args(exec_file):
+                yield test_tuple
 
 
-    def exec_test(self, test_file, test, clean_env=False):
-        keys = '"--filter={test}"'.format(test=test)
-        if clean_env:
-            tmpdir = tempfile.mkdtemp()
-            keys = '{old_key} "--data_path={tmpdir}"'.format(old_key=keys, tmpdir=tmpdir)
-            logging.debug("Temp dir: {tmpdir}".format(tmpdir=tmpdir))
-        else:
-            keys = "{old_key}{resource_path}".format(old_key=keys, resource_path=self.key_postfix)
-            logging.debug("Setting user_resource_path and data_path to {resource_path}".format(resource_path=self.key_postfix))
+    def set_instance_vars_from_options(self, options):
+        self.workspace_path = options.folder
+        interim_runlist = list()
+        for opt in options.runlist:
+            interim_runlist.extend(map(str.strip, opt.split(",")))
 
-        self.log_start_finish(test_file, keys, start=True)
-        out, err, result = self.get_tests_from_exec_file(test_file, keys)
-        self.log_start_finish(test_file, keys, finish=True)
+        tests_on_disk_list = tests_on_disk(self.workspace_path)
+        self.runlist = filter(lambda x: x in tests_on_disk_list, interim_runlist)
 
-        if clean_env:
-            try:
-                shutil.rmtree(tmpdir)
-            except:
-                logging.error("Failed to remove tempdir {tmpdir}".format(tmpdir=tmpdir))
+        self.params[RESOURCE_PATH_KEY] = options.user_resource_path
+        self.params[DATA_PATH_KEY] = options.data_path
 
-        with self.file_lock:
-            self.file.write("BEGIN: {file}\n".format(file=test_file))
-            self.file.write(str(err))
-            self.file.write("\nEND: {file} | result: {res}\n\n".format(file=test_file, res=result))
-            self.file.flush()
 
+    def process_cli(self):
+        parser = self.prepare_cli_parser()
+        (options, args) = parser.parse_args()
 
-    def get_tests_from_exec_file(self, test, keys):
-        spell = ["{test} {keys}".format(test=path.join(self.workspace_path, test), keys=keys)]
-        logging.debug(">> {spell}".format(spell=spell))
+        if not options.runlist:
```

Comment:
I think you can redone this with argparse accumulators and without manually exiting by Exception